zkir/Dialect/Field/Conversions/FieldToModArith/Extension/CubicExtensionField.cpp

```diff
@@ -16,10 +16,72 @@ limitations under the License.
 #include "zkir/Dialect/Field/Conversions/FieldToModArith/Extension/CubicExtensionField.h"
 
 #include "zkir/Dialect/Field/Conversions/FieldToModArith/ConversionUtils.h"
+#include "zkir/Dialect/Field/IR/FieldAttributes.h"
 #include "zkir/Dialect/ModArith/IR/ModArithOps.h"
+#include "zkir/Utils/APIntUtils.h"
 
 namespace mlir::zkir::field {
 
+CubicExtensionField::CubicExtensionField(ImplicitLocOpBuilder &b,
+                                         ExtensionFieldTypeInterface type,
+                                         const TypeConverter *converter)
+    : ExtensionField(b, type, converter) {
+  // Get the modulus from the base field type.
+  auto baseField = cast<PrimeFieldType>(type.getBaseFieldType());
+  llvm::APInt modulus = baseField.getModulus().getValue();
+  auto convertedType = converter->convertType(baseField);
+  unsigned bitWidth = modulus.getBitWidth();
+
+  // Zero constant for Frobenius map.
+  zero = b.create<mod_arith::ConstantOp>(
+      convertedType,
+      IntegerAttr::get(baseField.getStorageType(), llvm::APInt(bitWidth, 0)));
+
+  // Precompute Frobenius coefficients.
+  // For u³ = ξ, we have uᵏ⁽ᵖᵉ⁾ = ξᵏ⁽ᵖᵉ⁾/³ * uᵏ⁽ᵖᵉ⁾ ᵐᵒᵈ ³
+  auto nonResidueAttr = cast<PrimeFieldAttr>(type.getNonResidue());
+  llvm::APInt xi = nonResidueAttr.getValue().getValue();
+  if (xi.getBitWidth() < bitWidth) {
+    xi = xi.zext(bitWidth);
+  } else if (xi.getBitWidth() > bitWidth) {
+    xi = xi.trunc(bitWidth);
+  }
```

Comment:
do we need this?

Reply:
When does the bit width of non residue differ from the bit width of the modulus? Additionlaly, if it's possible, is it better to control the bit width of the non residue to ensure it is the same as the bit width of the modulus?

Reply:
oh.. PrimeFieldAttr::verify already checks that the value's bitwidth matches the modulus bitwidth.
I will remove this!